File: src/matchbook.c
#include "types.h"
#include <string.h>
#include <stdio.h>

void matchbook_init(MatchBook *book) {
    book->count = 0;
}

Match *matchbook_add(MatchBook *book, const Match *match) {
    if (book->count >= MAX_MATCHES) return NULL;
    book->matches[book->count] = *match;
    return &book->matches[book->count++];
}

void matchbook_print(const MatchBook *book) {
    printf("=== %u Matches ===\n\n", book->count);
    for (uint32_t i = 0; i < book->count; i++) {
        const Match *m = &book->matches[i];
        printf("%s vs %s (%s)\n", m->team1, m->team2, m->date);
        printf("\t %-25s  %.2f / %.2f  (%.0f%%)\n", m->team1, m->markets[TEAM1].bid, m->markets[TEAM1].ask, m->markets[TEAM1].prob * 100);
        printf("\t %-25s  %.2f / %.2f  (%.0f%%)\n", "Draw", m->markets[DRAW].bid, m->markets[DRAW].ask, m->markets[DRAW].prob * 100);
        printf("\t %-25s  %.2f / %.2f  (%.0f%%)\n", m->team2, m->markets[TEAM2].bid, m->markets[TEAM2].ask, m->markets[TEAM2].prob * 100);
        printf("\n");
    }
}

File: src/matchbook.rs
use crate::types::{Match, MatchBook, Outcome, MAX_MATCHES};

pub fn matchbook_init() -> MatchBook {
    MatchBook {
        matches: Vec::new(),
    }
}

pub fn matchbook_add(book: &mut MatchBook, m: Match) -> Option<&Match> {
    if book.matches.len() >= MAX_MATCHES {
        return None;
    }
    book.matches.push(m);
    book.matches.last()
}

pub fn matchbook_print(book: &MatchBook) {
    println!("=== {} Matches ===\n", book.matches.len());
    for m in &book.matches {
        println!("{} vs {} ({})", m.team1, m.team2, m.date);
        println!(
            "\t {:<25}  {:.2} / {:.2}  ({:.0}%)",
            m.team1,
            m.markets[Outcome::Team1.as_index()].bid,
            m.markets[Outcome::Team1.as_index()].ask,
            m.markets[Outcome::Team1.as_index()].prob * 100.0
        );
        println!(
            "\t {:<25}  {:.2} / {:.2}  ({:.0}%)",
            "Draw",
            m.markets[Outcome::Draw.as_index()].bid,
            m.markets[Outcome::Draw.as_index()].ask,
            m.markets[Outcome::Draw.as_index()].prob * 100.0
        );
        println!(
            "\t {:<25}  {:.2} / {:.2}  ({:.0}%)",
            m.team2,
            m.markets[Outcome::Team2.as_index()].bid,
            m.markets[Outcome::Team2.as_index()].ask,
            m.markets[Outcome::Team2.as_index()].prob * 100.0
        );
        println!();
    }
}
